File: apps/repo-worker/reference-ts/lib/refs.ts
import { HEX64 } from "./hex";

// Ref name grammar — mkit SPEC-REFS §3.
//
//   ref_name := segment ( '/' segment )*
//   segment  := char+
//   char     := ALNUM | '.' | '_' | '-'      (ALNUM = [0-9A-Za-z])
//
// Plus rejections: empty, leading '/', empty segment ('//' / trailing '/'),
// any segment == "." or "..", any byte in {0x00, '\\'}, any segment ending
// in ".lock", and a final segment equal to "HEAD".
//
// Validated at the transport boundary, as the spec requires.

const SEGMENT_CHAR = /^[0-9A-Za-z._-]+$/;

export function isValidRefName(name: string): boolean {
  if (name.length === 0) return false;
  if (name.startsWith("/")) return false;
  if (name.includes(" ") || name.includes("\\")) return false;

  const segments = name.split("/");
  for (const seg of segments) {
    if (seg.length === 0) return false; // empty segment (//, trailing/leading /)
    if (seg === "." || seg === "..") return false;
    if (!SEGMENT_CHAR.test(seg)) return false;
    if (seg.endsWith(".lock")) return false;
  }
  if (segments[segments.length - 1] === "HEAD") return false;
  return true;
}

/** A prefix for ListRefs is empty or a valid ref name, optionally trailing-'/'. */
export function isValidRefPrefix(prefix: string): boolean {
  if (prefix.length === 0) return true;
  const trimmed = prefix.endsWith("/") ? prefix.slice(0, -1) : prefix;
  return isValidRefName(trimmed);
}

/** A ref hash on the HTTP/JSON edge is 64-char lowercase hex; on the proto wire
 *  it is a raw 32-byte id. Both reduce to "32 bytes". */
export function isValidRefHashHex(value: string): boolean {
  return HEX64.test(value);
}

// ---------------------------------------------------------------------------
// CAS state machine — proto-aligned (RefExpectation: ANY / MISSING / MATCH).
//
// Mirrors mkit.repo.v1.RefExpectation (and mkit-rpc ssh.proto). The wire
// numbers are load-bearing.
// ---------------------------------------------------------------------------

export enum RefExpectation {
  UNSPECIFIED = 0,
  ANY = 1,
  MISSING = 2,
  MATCH = 3,
}

/**
 * Result of evaluating a CAS update. `conflict` distinguishes a precondition
 * failure (client should rebase + retry) from `invalid` (a malformed request,
 * e.g. UNSPECIFIED expectation or a non-empty expected_id where it must be
 * empty).
 */
export type CasDecision =
  | { kind: "committed" }
  | { kind: "conflict"; reason: "exists" | "missing" | "mismatch" }
  | { kind: "invalid"; reason: string };

/**
 * Pure CAS decision. Ids are compared as opaque strings — at the proto edge
 * pass lowercase-hex of the 32-byte ids; at a future Rust port pass the raw
 * bytes encoded consistently. `current` is null when the ref is absent.
 * `expected` is the MATCH target (ignored / required-empty otherwise).
 *
 *   ANY      → always commit (clobber); expected MUST be empty.
 *   MISSING  → commit iff current == null (else conflict "exists"); expected MUST be empty.
 *   MATCH    → commit iff current == expected (else conflict "missing"/"mismatch"); expected REQUIRED.
 */
export function evaluateCas(
  current: string | null,
  expectation: RefExpectation,
  expected: string | null,
): CasDecision {
  switch (expectation) {
    case RefExpectation.ANY:
      if (expected) return { kind: "invalid", reason: "expected_id must be empty for ANY" };
      return { kind: "committed" };

    case RefExpectation.MISSING:
      if (expected) return { kind: "invalid", reason: "expected_id must be empty for MISSING" };
      return current === null ? { kind: "committed" } : { kind: "conflict", reason: "exists" };

    case RefExpectation.MATCH:
      if (!expected) return { kind: "invalid", reason: "expected_id required for MATCH" };
      if (current === null) return { kind: "conflict", reason: "missing" };
      if (current !== expected) return { kind: "conflict", reason: "mismatch" };
      return { kind: "committed" };

    case RefExpectation.UNSPECIFIED:
    default:
      return { kind: "invalid", reason: "expectation is UNSPECIFIED (protocol error)" };
  }
}

File: apps/vcs-worker/src/refs.rs
// SPDX-License-Identifier: MIT OR Apache-2.0
//
// Ref-name/pack-id validation and the CAS decision state machine for
// UpdateRef / AdvanceRefs. Adapted from apps/repo-worker/src/refs.rs: this
// service has no "room" concept (one Worker deployment = one mkit
// repository, per SPEC-TRANSPORT-CONNECT §7.1), so the room allow-list is
// dropped; everything else (ref-name grammar delegation, the pure CAS state
// machine) is unchanged.

/// Validate a ref name against the mkit SPEC-REFS §3 grammar. Delegates to
/// the canonical [`mkit_core::refs::validate_ref_name`] so the worker and the
/// core crate can never desync on the grammar.
#[must_use]
pub fn is_valid_ref_name(name: &str) -> bool {
    mkit_core::refs::validate_ref_name(name)
}

/// A ListRefs prefix is empty, or a valid ref name (optionally trailing '/').
/// Delegates to the canonical [`mkit_core::refs::validate_ref_prefix`].
#[must_use]
pub fn is_valid_ref_prefix(prefix: &str) -> bool {
    mkit_core::refs::validate_ref_prefix(prefix)
}

/// CAS expectation, proto-aligned with `mkit.transport.v1.RefExpectation`.
/// Wire numbers are load-bearing (ANY=1, MISSING=2, MATCH=3; UNSPECIFIED=0) —
/// pinned byte-for-byte against `mkit.rpc.v1.ssh.RefExpectation` /
/// `mkit.repo.v1.RefExpectation` (see transport.proto's header comment).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RefExpectation {
    Unspecified = 0,
    Any = 1,
    Missing = 2,
    Match = 3,
}

impl RefExpectation {
    /// Map a raw proto enum wire number to the CAS expectation. Unknown
    /// numbers collapse to `Unspecified` (a protocol error downstream).
    #[must_use]
    pub fn from_wire(n: i32) -> Self {
        match n {
            1 => Self::Any,
            2 => Self::Missing,
            3 => Self::Match,
            _ => Self::Unspecified,
        }
    }
}

/// The reason a CAS update could not commit.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ConflictReason {
    /// MISSING expectation but the ref already exists.
    Exists,
    /// MATCH expectation but the ref is absent.
    Missing,
    /// MATCH expectation but the current value differs from `expected`.
    Mismatch,
}

/// Outcome of evaluating a CAS update — a pure decision. `Invalid` is a
/// malformed request (protocol error, SPEC-TRANSPORT-CONNECT §3: a
/// conforming server MUST reject `REF_EXPECTATION_UNSPECIFIED` with
/// `invalid_argument`); `Conflict` is a precondition failure the client can
/// rebase + retry (`failed_precondition`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CasDecision {
    Committed,
    Conflict(ConflictReason),
    Invalid(&'static str),
}

/// Pure CAS decision. `current` is the ref's present value (None = absent);
/// `expected` is the MATCH target (must be None for ANY/MISSING). Ids are
/// compared as opaque byte slices.
///
///   ANY      -> always commit (clobber); expected MUST be empty.
///   MISSING  -> commit iff current is None (else conflict Exists); expected MUST be empty.
///   MATCH    -> commit iff current == expected (else Missing/Mismatch); expected REQUIRED.
#[must_use]
pub fn evaluate_cas(
    current: Option<&[u8]>,
    expectation: RefExpectation,
    expected: Option<&[u8]>,
) -> CasDecision {
    match expectation {
        RefExpectation::Any => {
            if expected.is_some() {
                return CasDecision::Invalid("expected_id must be empty for ANY");
            }
            CasDecision::Committed
        }
        RefExpectation::Missing => {
            if expected.is_some() {
                return CasDecision::Invalid("expected_id must be empty for MISSING");
            }
            match current {
                None => CasDecision::Committed,
                Some(_) => CasDecision::Conflict(ConflictReason::Exists),
            }
        }
        RefExpectation::Match => {
            let Some(expected) = expected else {
                return CasDecision::Invalid("expected_id required for MATCH");
            };
            match current {
                None => CasDecision::Conflict(ConflictReason::Missing),
                Some(cur) if cur != expected => CasDecision::Conflict(ConflictReason::Mismatch),
                Some(_) => CasDecision::Committed,
            }
        }
        RefExpectation::Unspecified => {
            CasDecision::Invalid("expectation is UNSPECIFIED (protocol error)")
        }
    }
}

/// A raw 32-byte BLAKE3 pack digest (proto wire form for `pack_id` /
/// `object_id` / `new_id` / `expected_id` fields carrying a MATCH target).
#[must_use]
pub fn is_valid_digest(bytes: &[u8]) -> bool {
    bytes.len() == 32
}

#[cfg(test)]
mod tests {
    use super::*;

    const ID_A: &[u8] = &[0xaa; 32];
    const ID_B: &[u8] = &[0xbb; 32];

    #[test]
    fn any_clobbers() {
        assert_eq!(
            evaluate_cas(Some(ID_A), RefExpectation::Any, None),
            CasDecision::Committed
        );
        assert_eq!(
            evaluate_cas(None, RefExpectation::Any, None),
            CasDecision::Committed
        );
        assert!(matches!(
            evaluate_cas(Some(ID_A), RefExpectation::Any, Some(ID_A)),
            CasDecision::Invalid(_)
        ));
    }

    #[test]
    fn missing_create_only() {
        assert_eq!(
            evaluate_cas(None, RefExpectation::Missing, None),
            CasDecision::Committed
        );
        assert_eq!(
            evaluate_cas(Some(ID_A), RefExpectation::Missing, None),
            CasDecision::Conflict(ConflictReason::Exists)
        );
        assert!(matches!(
            evaluate_cas(None, RefExpectation::Missing, Some(ID_A)),
            CasDecision::Invalid(_)
        ));
    }

    #[test]
    fn match_cas() {
        assert_eq!(
            evaluate_cas(Some(ID_A), RefExpectation::Match, Some(ID_A)),
            CasDecision::Committed
        );
        assert_eq!(
            evaluate_cas(Some(ID_B), RefExpectation::Match, Some(ID_A)),
            CasDecision::Conflict(ConflictReason::Mismatch)
        );
        assert_eq!(
            evaluate_cas(None, RefExpectation::Match, Some(ID_A)),
            CasDecision::Conflict(ConflictReason::Missing)
        );
        assert!(matches!(
            evaluate_cas(Some(ID_A), RefExpectation::Match, None),
            CasDecision::Invalid(_)
        ));
    }

    #[test]
    fn unspecified_is_protocol_error() {
        assert!(matches!(
            evaluate_cas(None, RefExpectation::Unspecified, None),
            CasDecision::Invalid(_)
        ));
    }

    #[test]
    fn from_wire_numbers_match_proto() {
        assert_eq!(RefExpectation::from_wire(1), RefExpectation::Any);
        assert_eq!(RefExpectation::from_wire(2), RefExpectation::Missing);
        assert_eq!(RefExpectation::from_wire(3), RefExpectation::Match);
        assert_eq!(RefExpectation::from_wire(0), RefExpectation::Unspecified);
        assert_eq!(RefExpectation::from_wire(99), RefExpectation::Unspecified);
    }

    #[test]
    fn digest_length() {
        assert!(is_valid_digest(&[0u8; 32]));
        assert!(!is_valid_digest(&[0u8; 31]));
        assert!(!is_valid_digest(&[]));
    }
}
